From 3082a5e092e539f863f254ca30fe6da7c23a44a6 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Fri, 15 Dec 2023 17:25:00 +0100
Subject: [PATCH 0110/1695] drivers: stm32_iwdg: provide timeout range

Implement watchdog service init handler that is needed by U-Boot
to get min/max timeout range.

This change was picked from https://github.com/OP-TEE/optee_os/pull/6556.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 6a3b16a9c125a14825510d31c496acc7cedf723a drivers: stm32_iwdg: provide timeout range]
Change-Id: Id4b1922cacace33e1268f4f2e81091eb27e29587
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/drivers/stm32_iwdg.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/core/drivers/stm32_iwdg.c b/core/drivers/stm32_iwdg.c
index 35f700e23..d91882241 100644
--- a/core/drivers/stm32_iwdg.c
+++ b/core/drivers/stm32_iwdg.c
@@ -188,6 +188,23 @@ static struct stm32_iwdg_device *wdt_chip_to_iwdg(struct wdt_chip *chip)
 	return container_of(chip, struct stm32_iwdg_device, wdt_chip);
 }
 
+static TEE_Result iwdg_wdt_init(struct wdt_chip *chip,
+				unsigned long *min_timeout,
+				unsigned long *max_timeout)
+{
+	struct stm32_iwdg_device *iwdg = wdt_chip_to_iwdg(chip);
+	unsigned long rate = clk_get_rate(iwdg->clk_lsi);
+
+	if (!rate)
+		return TEE_ERROR_GENERIC;
+
+	/* Be safe and expect any counter to be above 2 */
+	*min_timeout = 3 * IWDG_PRESCALER_256 / rate;
+	*max_timeout = (IWDG_CNT_MASK + 1) * IWDG_PRESCALER_256 / rate;
+
+	return TEE_SUCCESS;
+}
+
 static void iwdg_wdt_start(struct wdt_chip *chip)
 {
 	struct stm32_iwdg_device *iwdg = wdt_chip_to_iwdg(chip);
@@ -227,6 +244,7 @@ static TEE_Result iwdg_wdt_set_timeout(struct wdt_chip *chip,
 }
 
 static const struct wdt_ops stm32_iwdg_ops = {
+	.init = iwdg_wdt_init,
 	.start = iwdg_wdt_start,
 	.ping = iwdg_wdt_refresh,
 	.set_timeout = iwdg_wdt_set_timeout,
-- 
2.47.2

